xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1125 @@
+}
+
+// TestSuccessfulFirstUpdate tests a basic scenario for the Graceful Switch Load
+// Balancer, where it is setup with a balancer which should populate the Current
+// Load Balancer. Any ClientConn updates should then be forwarded to this
+// Current Load Balancer.
+func (s) TestSuccessfulFirstUpdate(t *testing.T) {
+	_, gsb := setup(t)
+	if err := gsb.SwitchTo(mockBalancerBuilder1{}); err != nil {
+		t.Fatalf("Balancer.SwitchTo failed with error: %v", err)
+	}
+	if gsb.balancerCurrent == nil {
+		t.Fatal("balancerCurrent was not built out when a correct SwitchTo() call should've triggered the balancer to build")
+	}
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to the Current Load Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	// Updating ClientConnState should forward the update exactly as is to the
+	// current balancer.
+	if err := gsb.UpdateClientConnState(ccs); err != nil {
+		t.Fatalf("Balancer.UpdateClientConnState(%v) failed with error: %v", ccs, err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+}
+
+// TestTwoBalancersSameType tests the scenario where there is a graceful switch
+// load balancer setup with a current and pending Load Balancer of the same
+// type. Any ClientConn update should be forwarded to the current if no pending,
+// and the only the pending if the graceful switch balancer contains both a
+// current and pending. The pending Load Balancer should also swap into current
+// whenever it updates with a connectivity state other than CONNECTING.
+func (s) TestTwoBalancersSameType(t *testing.T) {
+	tcc, gsb := setup(t)
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to either the Current or Pending Load
+	// Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	gsb.SwitchTo(mockBalancerBuilder1{})
+	gsb.UpdateClientConnState(ccs)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+
+	// The current balancer reporting READY should cause this state
+	// to be forwarded to the Client Conn.
+	gsb.balancerCurrent.Balancer.(*mockBalancer).updateState(balancer.State{
+		ConnectivityState: connectivity.Ready,
+		Picker:            &neverErrPicker{},
+	})
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
```

Comment:
`part UpdateState call` what does that supposed to mean?

Reply:
I'm trying to remember, but what it's supposed to mean isn't coming to me. Went ahead and switched this to "timeout while waiting for a UpdateState call on the ClientConn"